### PR TITLE
fix: fix unsupported webp for safari #2551

### LINKED ISSUE #2551: [bug]: responses with 404 code should not be cached
**Describe the bug**
Hi guys. I see we use `apicache` connected with hastly to optimize and cache images. We face with the issue caching of 404 responses on media proxy. I mean when something is wrong with the media server and proxy is not able to create image then we get 404, but after when media server is back and images are available, then cache is still valid and we see 404 error page still. 

**The second issue is about webp support and caching:**

When you make a deploy and you open the website on Chrome first (which supports webp) you will be able to see webp images. Then if you open the same page on the safari (which doesn't supports webp) you won't be able to see images because they were cached as webp with the first hit ( on chrome) 

**To reproduce**
1. When you use image optimize onboard
2. Open some image which doesn't exist on your media server
3. You should see 404 code response and cache headers
4. Create this missing image on backend server
5. You should see 404 code response

**Expected behavior**
Images with 404 and error headers should not be cached. 

**Screenshots**
Zrzut ekranu 2020-07-13 o 10 38 12

**Possible solutions**
My solution is simple according to documentation of apicache, but should we done it here or maybe it should be fixed on server side, because this solution is transfering the headers ?

```js
  cacheMiddleware = cache(cacheExpires, hastily.hasSupportedExtension, {
            debug: cacheDebug,
            statusCodes: {
                include: [200, 304]
              },
        });
```

**EDIT**

include, exclude option in statusCode brings no results until we changed second parameter of apicache instance this is 
hastily.hasSupportedExtension:

```(req, res) => hastily.hasSupportedExtension(req) && res.statusCode === 200```

Can I change this setup from my project level - webpackConfig or somewhere ? Because now I have to modify packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js to make it works. 

**Please let us know what packages this bug is in regards to:**
- [ ] `venia-concept`
- [ ] `venia-ui`
- [x] `pwa-buildpack`
- [ ] `peregrine`
- [ ] `pwa-devdocs`
- [ ] `upward-js`
- [ ] `upward-spec`
- [ ] `create-pwa`

### PR DESCRIPTION
**Pull request by mercs600**
## Description

resolves #2551 and #2707. The main issue for us was that image requests with 404 response were cached. The second issue is when you open images on chrome which supports webp and after that you  try to open the same image request on safari - where webp is not suported  - you will get broken request. This tweak is working for us on production. 

## Related Issue
#2551 and #2707.

### Verification Steps
Go with scenario from #2551 and #2707.

@zetlen

### CODE REVIEW

File: packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js [hunk 1]
```diff
@@ -29,9 +37,20 @@ function addImgOptMiddleware(app, config) {
...
-            debug: cacheDebug
-        });
+        cacheMiddleware = cache(
+            cacheExpires,
+            (req, res) =>
+                hastily.hasSupportedExtension(req) && res.statusCode === 200,
+            {
+                debug: cacheDebug,
+                appendKey: req =>
+                    IsWebpSupported(req) ? 'supported' : 'regular',
```
Comment (tjwiebell) [inline packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js:47]: This definitely solves the problem, but I'm concerned why it is needed. The `Vary` header in the response contains the `accept` header, and this lib doesn't seem to be respecting it. If we proceed with this workaround, I would probably just want `appendKey: req => req.headers.accept` while we investigate why this lib isn't respecting the `Vary` header.

**Review by tjwiebell — DISMISSED** [review]
Great work! This solves both problems and is a huge improvement for anyone using this middleware in production. I have some suggestions to make the cache key less opinionated, but I'm open to discussing.

**Review by revanth0212 — APPROVED** [review]
Basically, we will be checking for the status code before caching so we don't cache invalid responses. Nice work.

**Comment by dpatil-magento** [comment]
QA Approved.

File: packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js [hunk 2]
```diff
@@ -29,9 +29,15 @@ function addImgOptMiddleware(app, config) {
...
-        cacheMiddleware = cache(cacheExpires, hastily.hasSupportedExtension, {
-            debug: cacheDebug
-        });
+        cacheMiddleware = cache(
+            cacheExpires,
+            (req, res) =>
+                hastily.hasSupportedExtension(req) && res.statusCode === 200,
+            {
+                debug: cacheDebug,
+                appendKey: req => req.get('accept')
```
Comment (mercs600) [inline packages/pwa-buildpack/lib/Utilities/addImgOptMiddleware.js:38]: @tjwiebell it will be enough to support or not webp images for safari ?

**Pull request merged**